### PR TITLE
Add Rounded box gizmos

### PR DESCRIPTION
**Pull request by solis-lumine-vorago**
# Objective

- Implement rounded cuboids and rectangles, suggestion of #9400 

## Solution

- Added `Gizmos::rounded_cuboid`, `Gizmos::rounded_rect` and `Gizmos::rounded_rect_2d`.
- All of these return builders that allow configuring of the corner/edge radius using `.corner_radius(...)` or `.edge_radius(...)` as well as the line segments of each arc using `.arc_segments(...)`.

---

## Changelog

- Added a new `rounded_box` module to `bevy_gizmos` containing all of the above methods and builders.
- Updated the examples `2d_gizmos` and `3d_gizmos`

## Additional information

The 3d example now looks like this:

Screenshot 2024-02-28 at 01 47 28

And this is the updated 2d example showcasing negative corner radius:

Screenshot 2024-02-28 at 01 59 37
Screenshot 2024-02-28 at 01 59 47

### CODE REVIEW

File: crates/bevy_gizmos/src/rounded_box.rs [hunk 1]
```diff
@@ -0,0 +1,379 @@
+//! Additional [`Gizmos`] Functions -- Rounded cuboids and rectangles
+//!
+//! Includes the implementation of[`Gizmos::rounded_rect`], [`Gizmos::rounded_rect_2d`] and [`Gizmos::rounded_cuboid`].
```
Comment (JMS55) [inline crates/bevy_gizmos/src/rounded_box.rs:3]:
```suggestion
//! Includes the implementation of [`Gizmos::rounded_rect`], [`Gizmos::rounded_rect_2d`] and [`Gizmos::rounded_cuboid`].
```

File: crates/bevy_gizmos/src/rounded_box.rs [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+        // Calculate inner and outer half size and ensure that the edge_radius is <= any half_length
+        let outer_half_size = self.size.abs() / 2.0;
+        let inner_half_size =
+            (outer_half_size - Vec3::splat(config.corner_radius.abs())).max(Vec3::ZERO);
+        let mut edge_radius = (outer_half_size - inner_half_size).min_element();
+        let inner_half_size = outer_half_size - Vec3::splat(edge_radius);
+        edge_radius *= config.corner_radius.signum();
+
+        // Handle cases where the rounded cuboid collapses into simpler shapes
+        if edge_radius == 0. {
```
Comment (JMS55) [inline crates/bevy_gizmos/src/rounded_box.rs:161]:
```suggestion
        if edge_radius == 0.0 {
```

**Review by JMS55 — APPROVED** [review]
Never reviewed gizmo stuff before, but lgtm.

File: crates/bevy_gizmos/src/rounded_box.rs [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    /// # bevy_ecs::system::assert_is_system(system);
+    /// ```
+    pub fn rounded_rect(
+        &mut self,
+        position: Vec3,
+        rotation: Quat,
+        size: Vec2,
+        color: impl Into<Color>,
+    ) -> RoundedRectBuilder<'_, 'w, 's, T> {
+        let corner_radius = size.min_element() / 10.0;
```
Comment (jgayfer) [inline crates/bevy_gizmos/src/rounded_box.rs:252]: Maybe we should pull the default of `10.0` into a constant? Both to make it clear where it's coming from, and to split up the need to tweak the implementation if in the future the default needs to change.

I'm not 100% sure if we could also reference it from the docs, but that could help us keep a single definition for the default.

EDIT: Just saw we're doing this for `DEFAULT_ARC_SEGMENTS` already 👍 Could be handy to give corner radius the same treatment.

**Review by jgayfer — APPROVED** [review]
Not a Gizmo export by any means, but ran the examples and overall looks good to me.

Could be nice to pull that default into a constant, but not blocking.

File: crates/bevy_gizmos/src/rounded_box.rs (hunk 3 above)
Comment (solis-lumine-vorago) [reply crates/bevy_gizmos/src/rounded_box.rs]: Good suggestion :)

**Review by jgayfer — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1315 if you'd like to help out.

**Pull request merged**